### PR TITLE
Fix: Context files are not showing up in the chat

### LINKED ISSUE #328: Context files are not showing up in the chat

### PR DESCRIPTION
**Pull request by mkondratek**
## Test plan
- chat and see the context files (try clickable links)

### CODE REVIEW

File: src/main/kotlin/com/sourcegraph/cody/chat/ui/MessagesPanel.kt [hunk 1]
```diff
@@ -24,10 +26,14 @@ class MessagesPanel(private val project: Project) :
...
 
     if (componentCount > 0) {
-      val lastPanel = components.last() as? JPanel
-      val lastMessage = lastPanel?.getComponent(0) as? SingleMessagePanel
-      if (message.id == lastMessage?.getMessageId()) {
-        lastMessage.updateContentWith(message)
+      val messageToUpdate =
+          components
+              .mapNotNull { (it as? JPanel) }
+              .filter { it.componentCount > 0 }
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/chat/ui/MessagesPanel.kt]: unrel fix - previously `getComponent(0)` could throw index-out-of-bound

File: src/main/kotlin/com/sourcegraph/cody/chat/ui/MessagesPanel.kt [hunk 2]
```diff
@@ -24,10 +26,14 @@ class MessagesPanel(private val project: Project) :
...
+              .map { it.getComponent(0) as? SingleMessagePanel }
+              .find { it?.getMessageId() == message.id }
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/chat/ui/MessagesPanel.kt]: unrel fix - `find` instead of `last` (so we can update the message wherever it is)

File: src/main/kotlin/com/sourcegraph/cody/chat/ui/MessagesPanel.kt [hunk 3]
```diff
@@ -43,6 +49,32 @@ class MessagesPanel(private val project: Project) :
     repaint()
   }
 
+  @RequiresEdt
+  @Synchronized
+  fun addOrUpdateContext(message: ChatMessage) {
+    removeBlinkingCursor()
+    createUsedContextFilesPanel(message)?.let(::addComponentToChat)
+    add(BlinkingCursorComponent.instance)
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/chat/ui/MessagesPanel.kt]: I remove the blinking cursor and re-add it just after the context files panel. We receive context files much faster than the actual response. Trust me, it looks good now.

File: src/main/kotlin/com/sourcegraph/cody/chat/ChatSession.kt [hunk 4]
```diff
@@ -59,6 +59,8 @@ private constructor(private val project: Project, newSessionId: CompletableFutur
 
   private val logger = LoggerFactory.getLogger(ChatSession::class.java)
 
+  private val theMostRecentContextFilesMessageId = AtomicReference<UUID>(UUID.randomUUID())
```
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/chat/ChatSession.kt:62]: I don't like that state variable. It is used for updating single message context but it pollutes whole class. We should think about:
a) removing it completely somehow
b) if not possible then maybe moving to single message - then you don't need fake start UUID, just a boolean

File: src/main/kotlin/com/sourcegraph/cody/chat/ChatSession.kt [hunk 5]
```diff
@@ -179,6 +201,25 @@ private constructor(private val project: Project, newSessionId: CompletableFutur
...
+  @RequiresEdt
+  private fun addAssistantUsedContextToChat(message: ChatMessage) {
+    synchronized(this) {
+      val findResult = messages.indexOfLast { it.id == message.id }
+      if (findResult >= 0) {
+        messages[findResult] = message
+
+        // The first response does not include the context files. We must check
+        // for the second version of it when it's updated with the proper data.
+        val isFirstUpdate = theMostRecentContextFilesMessageId.getAndSet(message.id) != message.id
```
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/chat/ChatSession.kt:213]: You don't need that class-wide `theMostRecentContextFilesMessageId` variable.
You can just check if last message in `messages` have the same UUID. If it's the same, it's an update.

File: src/main/kotlin/com/sourcegraph/cody/chat/ChatSession.kt [hunk 6]
```diff
@@ -179,6 +201,25 @@ private constructor(private val project: Project, newSessionId: CompletableFutur
     }
   }
 
+  @RequiresEdt
+  private fun addAssistantUsedContextToChat(message: ChatMessage) {
+    synchronized(this) {
+      val findResult = messages.indexOfLast { it.id == message.id }
```
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/chat/ChatSession.kt:207]: I think more idiomatic for kotlin would be using find and check if not null, but it's a nitpick.

File: src/main/kotlin/com/sourcegraph/cody/chat/ChatSession.kt [hunk 7]
```diff
@@ -155,8 +157,28 @@ private constructor(private val project: Project, newSessionId: CompletableFutur
               extensionMessage.isMessageInProgress == false) {
             getCancellationToken().dispose()
           } else {
-            if (lastMessage?.text != null && extensionMessage.chatID != null) {
-              addAssistantResponseToChat(lastMessage.text, lastMessage.displayText)
+
+            val penultimateMessage = extensionMessage.messages?.dropLast(1)?.lastOrNull()
```
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/chat/ChatSession.kt]: I do not even know what `penultimate` means :) Can you use some other name?

File: src/main/kotlin/com/sourcegraph/cody/chat/ChatSession.kt [hunk 8]
```diff
@@ -155,8 +157,28 @@ private constructor(private val project: Project, newSessionId: CompletableFutur
...
+            if (extensionMessage.chatID != null) {
```
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/chat/ChatSession.kt]: We should not care about `chatId` anywhere in our logic. Why is that needed?

File: src/main/kotlin/com/sourcegraph/cody/chat/ChatSession.kt [hunk 9]
```diff
@@ -155,8 +157,28 @@ private constructor(private val project: Project, newSessionId: CompletableFutur
...
+              if (penultimateMessage != null) {
+                val messageId =
+                    UUID.nameUUIDFromBytes(
+                        (extensionMessage.messages.count() - 1).toString().toByteArray())
```
Comment (pkukielka) [inline src/main/kotlin/com/sourcegraph/cody/chat/ChatSession.kt]: Why `-1`? It is different formula than used in `addAssistantResponseToChat`, this will cause us issues.

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/ChatMessage.kt [hunk 10]
```diff
@@ -32,9 +31,6 @@ data class ChatMessage(
     val displayText: String? = null,
     val contextFiles: List<ContextFile>? = null,
     val error: ChatError? = null,
-    // Internal ID used for identifying updates of the message
-    // All partial messages which are part of the same response are required to have the same ID
-    val id: UUID = UUID.randomUUID()
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/agent/protocol/ChatMessage.kt:37]: note: this is great 🚀

File: src/main/kotlin/com/sourcegraph/cody/agent/protocol/Speaker.kt [hunk 11]
```diff
@@ -1,12 +1,13 @@
...
 import com.google.gson.annotations.SerializedName
 
 enum class Speaker(val speaker: String) {
   @SerializedName("human") HUMAN("human"),
   @SerializedName("assistant") ASSISTANT("assistant")
 }
 
-val SpeakerSerializer = JsonSerializer<Speaker> { src, _, _ -> JsonPrimitive(src.speaker) }
+val speakerDeserializer = JsonDeserializer { src, _, _ ->
+  Speaker.values().find { it.speaker == src.asString }
```
Comment (mkondratek) [inline src/main/kotlin/com/sourcegraph/cody/agent/protocol/Speaker.kt:12]: still, not the best but rather simple and working - I assume our problems with this field come from the fact that we mix serialization and inheritance (here in the client and in the agent)

**Comment by mkondratek** [comment]
closing in favor of https://github.com/sourcegraph/jetbrains/pull/520
